protocol/consistencychecks.go

```diff
@@ -0,0 +1,265 @@
+	if str.Epoch > cc.CurrentEpoch {
+		cc.CurrentEpoch++
+		cc.SavedSTR = str.Signature
+	}
+	return
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProofs(dfs *DirectoryProofs,
+	uname string, key []byte) (int, error) {
+	// TODO: fix me
+	return invalidProof, nil
+}
+
+func (cc *ConsistencyChecks) verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (int, error) {
+	proofType := proofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return invalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return invalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return invalidProof, ErrorBadCommitment
+		}
+		proofType = proofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
+		(str.Epoch == cc.CurrentEpoch+1 && str.VerifyHashChain(cc.SavedSTR)) {
+		return nil
+	}
+	return ErrorBadSTR
+}
+
+// verifyReturnedPromise verifies the returned TB
+// included in the directory's response.
+func (cc *ConsistencyChecks) verifyReturnedPromise(tb *m.TemporaryBinding,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if cc.isUseTBs && tb != nil {
+		// verify TB's Signature
+		if !cc.signKey.Verify(tb.Serialize(str.Signature), tb.Signature) {
+			return ErrorBadSignature
+		}
+
+		// verify TB's VRF index
+		if !bytes.Equal(tb.Index, ap.LookupIndex) {
+			return ErrorBadIndex
+		}
+
+		cc.TBs = append(cc.TBs, tb)
+	}
+
+	return nil
+}
+
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if cc.isUseTBs {
+		backed := cc.TBs[:0]
+		for _, tb := range cc.TBs {
+			if cc.CurrentEpoch+1 == str.Epoch {
+				if !tb.Verify(ap) {
+					return ErrorBrokenPromise
+				}
+			} else if cc.CurrentEpoch == str.Epoch {
+				// keep current epoch's returned promises
+				// for future verifications
+				backed = append(backed, tb)
+			}
+		}
+
+		// clear all issued promises since they have been verified
+		// or the client has missed some epochs
+		cc.TBs = backed
+	}
+	return nil
+}
+
+func (cc *ConsistencyChecks) verifyProofTypeWithTB(statusCode ErrorCode, proofType int) ErrorCode {
```

Comment:
I might be missing something, but I don't really understand the purpose of this function. Can you please clarify what this function is supposed to do? Also, I'm a bit misled by the name of this function since it doesn't actually involve TBs in anyway.